quisp/modules/QRSA/RuleEngine/RuleEngine.cc

```diff
@@ -994,23 +986,27 @@ void RuleEngine::updateResources_EntanglementSwapping(swapping_result swapr) {
   }
 
   if (qubit->entangled_partner == nullptr && qubit->Density_Matrix_Collapsed(0, 0).real() == -111 && !qubit->no_density_matrix_nullptr_entangled_partner_ok) {
-    std::cout << qubit << ", node[" << qubit->node_address << "] from qnic[" << qubit->qnic_index << "]\n";
-    // std::cout<<(bool)(qubit->entangled_partner==nullptr)<<" Entangled if ("<<false<<")\n";
-    // std::cout<<qubit->Density_Matrix_Collapsed<<"\n";
-    EV << "This is node" << qubit->entangled_partner << "\n";
-    error("RuleEngine. Ebit succeed. but wrong");
+    // std::cout << qubit << ", node[" << qubit->node_address << "] from qnic[" << qubit->qnic_index << "]\n";
+    error("RuleEngine. Entanglement swapping went wrong");
   }
-  allResources[qnic_type][qnic_index].insert(std::make_pair(new_partner /*QNode IP address*/, qubit));
+  // first delete old record
+  for (auto it = allResources[qnic_type][qnic_index].begin(); it != allResources[qnic_type][qnic_index].end(); ++it) {
+    if (it->second == qubit) {
+      allResources[qnic_type][qnic_index].erase(it);
```

Comment:
DO NOT just erase it inside the loop. it breaks the structure and iterator itself.

```cpp

  auto &resource  = allResources[qnic_type][qnic_index];
  // first delete old record
  for (auto it = resource.begin(); it != resource.end(); ++it) {
    if (it->second == qubit) {
      resource.erase(it++);
    }
  }
```

Reply:
sorry ^ this snippet doesn't work well because it increments twice in one loop.
```cpp
  auto &resource  = allResources[qnic_type][qnic_index];
  auto it = resource.begin();
  // first delete old record
  while (it != resource.end()) {
    if (it->second == qubit) {
      resource.erase(it++);
    } else {
      it++;
    }
  }
```

Reply:
I believe you can use `resource.erase(it--);`

Reply:
naphann's way is more readable for me.

Reply:
OK, let me fix that.